### PR TITLE
Feature/openglfixes

### PR DESCRIPTION
**Pull request by petersteneteg**
Fix for qt context thread id
The Shader editor now has run butotn (Ctrl-R) to apply changes wth out writing to disk.
The Shader segments now uses more string_views.
The MeshShaderCache now avoids some extra copies.
Fixes a shader registration issue.

### CODE REVIEW

File: include/inviwo/core/util/constexprhash.h [hunk 1]
```diff
@@ -129,25 +130,20 @@ static uint64_t constexpr crc_table[] = {
...
+        crc = (crc >> 8) ^ detail::crc_table[(crc ^ c) & 0xff];
+    }
+    return crc ^ 0xffffffff;
 }
 
-constexpr uint64_t constexpr_hash(const std::string_view str) {
-    return detail::crc64(str.data(), str.length());
+template <size_t len>
+[[nodiscard]] constexpr std::uint64_t constexpr_hash(const char (&str)[len]) noexcept {
+    static_assert(len > 0, "len should include the null terminator");
```
Comment (martinfalk) [inline include/inviwo/core/util/constexprhash.h:145]: shouldn't you rather check `len>0 && str[len-1] == '\0'` ?

File: modules/opengl/include/modules/opengl/shader/shadersegment.h [hunk 2]
```diff
@@ -90,9 +94,9 @@ struct IVW_MODULE_OPENGL_API ShaderSegment {
...
 template <>
-struct hash<typename inviwo::ShaderSegment::Type> {
-    size_t operator()(typename inviwo::ShaderSegment::Type const& type) const {
-        return std::hash<std::string>{}(type.getString());
+struct hash<typename inviwo::ShaderSegment::Placeholder> {
+    size_t operator()(typename inviwo::ShaderSegment::Placeholder const& type) const {
+        return std::hash<std::string_view>{}(type.key);
     }
 };
 }  // namespace std
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/shader/shadersegment.h:102]: missing endline

File: modules/openglqt/src/shaderwidget.cpp [hunk 3]
```diff
@@ -81,12 +84,34 @@ ShaderWidget::ShaderWidget(ShaderObject* obj, QWidget* parent)
...
+        "update the string. The change will affect all shaders using the resource but will not be "
+        "persistent");
     save_->setShortcut(QKeySequence::Save);
     save_->setShortcutContext(Qt::WidgetWithChildrenShortcut);
     mainWindow->addAction(save_);
     connect(save_, &QAction::triggered, this, &ShaderWidget::save);
 
+    revert_ = toolBar->addAction(QIcon(":/svgicons/revert.svg"), tr("Revert"));
+    revert_->setToolTip("Revert changes");
+    revert_->setShortcutContext(Qt::WidgetWithChildrenShortcut);
```
Comment (martinfalk) [inline modules/openglqt/src/shaderwidget.cpp:109]: and the shortcut is?

File: src/qt/editor/inviwomainwindow.cpp [hunk 4]
```diff
@@ -836,49 +836,92 @@ void InviwoMainWindow::addActions() {
...
-                action->setCheckable(true);
-                action->setChecked(p->getProcessorWidget()->isVisible());
-                QObject::connect(action, &QAction::toggled, this,
-                                 [p](bool toggle) { p->getProcessorWidget()->setVisible(toggle); });
-            }
-        });
+        auto getFloatingDockWidgets = [](InviwoMainWindow* win) {
+            auto dockWidgets = util::copy_if(win->findChildren<InviwoDockWidget*>(),
+                                             [](const auto p) { return p->isFloating(); });
+            std::sort(dockWidgets.begin(), dockWidgets.end(), [](auto a, auto b) {
```
Comment (martinfalk) [inline src/qt/editor/inviwomainwindow.cpp:864]: what's the reason for sorting here?

**Review by martinfalk — APPROVED** [review]

File: src/qt/editor/inviwomainwindow.cpp (hunk 4 above)
Comment (petersteneteg) [reply src/qt/editor/inviwomainwindow.cpp:864]: to show them sorted in the menu

**Pull request merged**